cmd/private-org-peribolos-sync/main.go

```diff
@@ -0,0 +1,230 @@
+
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"k8s.io/test-infra/prow/config/org"
+	"k8s.io/test-infra/prow/config/secret"
+	"k8s.io/test-infra/prow/flagutil"
+	"k8s.io/test-infra/prow/github"
+
+	"github.com/openshift/ci-tools/pkg/api"
+	"github.com/openshift/ci-tools/pkg/config"
+	"github.com/openshift/ci-tools/pkg/promotion"
+)
+
+type gitHubClient interface {
+	GetRepo(owner, name string) (github.FullRepo, error)
+}
+
+type options struct {
+	peribolosConfig string
+	destOrg         string
+	releaseRepoPath string
+	whitelist       flagutil.Strings
+	github          flagutil.GitHubOptions
+
+	whitelistByOrg map[string]sets.String
+}
+
+func gatherOptions() options {
+	o := options{}
+	fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+
+	fs.StringVar(&o.peribolosConfig, "peribolos-config", "", "Peribolos configuration file")
+	fs.StringVar(&o.releaseRepoPath, "release-repo-path", "", "Path to a openshift/release repository directory")
+	fs.StringVar(&o.destOrg, "destination-org", "", "Destination name of the peribolos configuration organzation")
+
+	fs.Var(&o.whitelist, "whitelist", "One or more repositories that do not promote official images and need to be included")
+
+	o.github.AddFlags(fs)
+	fs.Parse(os.Args[1:])
+
+	return o
+}
+
+func validateOptions(o *options) error {
+	var validationErrors []error
+
+	if len(o.releaseRepoPath) == 0 {
+		validationErrors = append(validationErrors, errors.New("--release-repo-path is not specified"))
+	}
+	if len(o.peribolosConfig) == 0 {
+		validationErrors = append(validationErrors, errors.New("--peribolos-config is not specified"))
+	}
+	if len(o.destOrg) == 0 {
+		validationErrors = append(validationErrors, errors.New("--destination-org is not specified"))
+	}
+
+	if len(o.whitelist.Strings()) > 0 {
+		if err := validateWhitelist(o.whitelist.Strings()); err != nil {
+			validationErrors = append(validationErrors, err)
+		} else {
+			o.whitelistByOrg = getWhitelistByOrg(o.whitelist.Strings())
+		}
+	}
+
+	if err := o.github.Validate(false); err != nil {
+		validationErrors = append(validationErrors, err)
+	}
+
+	return kerrors.NewAggregate(validationErrors)
+}
+
+func validateWhitelist(whitelist []string) error {
+	for _, repo := range whitelist {
+		if len(strings.Split(repo, "/")) != 2 {
+			return fmt.Errorf("--whitelist: %s must be in org/repo format", repo)
+		}
+	}
+	return nil
+}
+
+func getWhitelistByOrg(whitelist []string) map[string]sets.String {
+	ret := make(map[string]sets.String)
+
+	for _, repo := range whitelist {
+		orgRepo := strings.Split(repo, "/")
+
+		orgName := orgRepo[0]
+		repoName := orgRepo[1]
+
+		ret[orgName] = ret[orgName].Union(sets.NewString(repoName))
+	}
+	return ret
+}
+
+func main() {
+	o := gatherOptions()
+	err := validateOptions(&o)
+	if err != nil {
+		logrus.WithError(err).Fatal("invalid options")
+	}
+	logger := logrus.WithField("destination-org", o.destOrg)
+
+	sigTerm := make(chan os.Signal, 1)
```

Comment:
Why? Just use `prow/interrupts`

Reply:
It seems that `prow/interrupts` library is for more complex activities. Here we just want to kill the program when it receives an interruption.

Reply:
Then don't do anything at all. Your program will exit when it gets `SIGINT` and will do so with a nonzero code.

Reply:
It seems that it is not exiting... I guess something is going in the `prow/github` package.

Reply:
So it seems that I have to use something like https://github.com/openshift/ci-tools/blob/master/cmd/repo-init/main.go#L89-L92 to actual exit, but this library is not meant to be used that way.

Reply:
Please look upstream for every single user of that library and copy how they use it :)

Reply:
The `repo-init` is special since it's interacting with users on the commandline

Reply:
I investigated this, and I can't find any reasonable way to use this library in this tool. I can use some insights :)

Reply:
With further investigation, I got it. Addressed.